### PR TITLE
Add "discard_writes" option to --pmem

### PR DESCRIPTION
**Pull request by rbradford**
Unfortunately there lacks the ability to mark a virtio-pmem device readonly however we instead get "discard-on-write" behaviour where changes are not persisted after a reboot.

### CODE REVIEW

**Comment by rbradford** [comment]
Going further I think I want `MAP_PRIVATE` for the readonly case.

File: vmm/src/device_manager.rs [hunk 1]
```diff
@@ -1383,12 +1383,16 @@ impl DeviceManager {
...
                     (0, false)
                 };
 
+                if pmem_cfg.readonly {
+                    custom_flags |= O_RDONLY;
+                }
+
                 let file = OpenOptions::new()
                     .read(true)
                     .write(true)
```
Comment (eryugey) [inline vmm/src/device_manager.rs]: Why not just set .write(false) here?

Comment (rbradford) [reply vmm/src/device_manager.rs]: @eryugey thanks for the idea I incorporated it.

File: vmm/src/device_manager.rs [hunk 2]
```diff
@@ -1405,12 +1405,17 @@ impl DeviceManager {
...
                         PROT_READ
                     } else {
                         PROT_READ | PROT_WRITE
                     },
-                    MAP_NORESERVE | MAP_SHARED,
+                    MAP_NORESERVE
+                        | if pmem_cfg.discard_writes {
+                            MAP_PRIVATE
+                        } else {
+                            MAP_SHARED
```
[the remark this answers is not in the document]
Comment (rbradford) [reply vmm/src/device_manager.rs]: From mmap(2)
```
             MAP_PRIVATE   
              Create  a private copy-on-write mapping.  Updates to the mapping
              are not visible to other processes mapping the  same  file,  and
              are  not carried through to the underlying file.  It is unspeci‐
              fied whether changes made to the file after the mmap() call  are
              visible in the mapped region.

```

File: vmm/src/device_manager.rs [hunk 3]
```diff
@@ -1391,7 +1391,7 @@ impl DeviceManager {
 
                 let file = OpenOptions::new()
                     .read(true)
-                    .write(true)
+                    .write(pmem_cfg.discard_writes)
```
[the remark this answers is not in the document]
Comment (rbradford) [reply vmm/src/device_manager.rs]: Oops, yes. Fixed.

File: src/main.rs [hunk 4]
```diff
@@ -173,7 +173,7 @@ fn create_app<'a, 'b>(
                 .help(
                     "Persistent memory parameters \
                      \"file=<backing_file_path>,size=<persistent_memory_size>,iommu=on|off,\
-                     mergeable=on|off\"",
+                     mergeable=on|off,readonly=on|off,\"",
```
[the remark this answers is not in the document]
Comment (rbradford) [reply src/main.rs]: Yup. And I even remembered to add it to the openapi spec.

**Pull request merged**